src/EditorFeatures/Core/Tagging/AbstractAsynchronousTaggerProvider.cs

```diff
@@ -91,6 +91,11 @@ internal abstract partial class AbstractAsynchronousTaggerProvider<TTag> where T
         /// </summary>
         protected virtual TaggerDelay AddedTagNotificationDelay => TaggerDelay.NearImmediate;
 
+        /// <summary>
+        /// Whether or not events from the <see cref="ITaggerEventSource"/> should cancel in-flight tag-computation.
+        /// </summary>
+        protected virtual bool CancelOnNewWork { get; }
```

Comment:
something i'd like us to do as well.  Syntactic compiler diagnostics are likely fast enough that we can just compute them all the way through without cancelling.  But things like analyzer diagnostics likely should be canceled on edit so that we're only computing the very latest info to show hte user.

Reply:
> But things like analyzer diagnostics likely should be canceled on edit so that we're only computing the very latest info to show hte user.

I think we probably _have_ to do this work before re-enabling pull diagnostics based tagging. Many feedback items from the corresponding 17.4 regression state that the delay in refreshing squiggles keeps increasing the longer you have the VS session up. This likely means lot of diagnostic computations are getting queued up before we get to the latest snapshot. The solution crawler based incremental analysis has a work queue which cancels prior work items on edit ensuring this logic. I believe we absolutely need to test squiggles refresh performance after long VS sessions when we re-enable pull diagnostics based tagging.

Reply:
note: the queues here are very different.  In tagging (and most IDE features) the 'queue' really is just a place where events are buffered.  Then, there is one active task that is given all the events that were around when the task starts running, and then the queue buffers more events until the next task run.  Each task is given *all* the events, so that it can compute smartly (e.g. ignoring events that are superceded by later events).  In tagging, there is no data in the events to begin with.  So really it's just one aggregate event that kicked off the currrent task, and then the next aggregate event which indicates if tagging is done, or if we should kick off the next task once the last one is done.

In no way do we get a queue that grows ever longer as the feature just processes one item in teh queue at a time.

--

So, what's the benefit of hte 'CancelOnNewWork' flag?  Well, it controls the behavior around those two tasks. Specifically, if the first task should run to completion before the second, or if we should try to cancel the first and then jump to the second asap.  So, it takes us from 2N cost to TimeSpentSoFar+N.  

That said, this seems like a pure win (at least for semantic tagging).  Better to not wait for the current pass to finish just to start the next one.